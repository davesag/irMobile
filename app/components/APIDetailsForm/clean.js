export const cleanKey = key => key.replace(/[^a-f0-9-]/gi, '')
export const cleanSecret = key => key.replace(/[^a-f0-9]/gi, '')
