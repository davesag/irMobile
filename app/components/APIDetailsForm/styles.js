import { Platform, StyleSheet } from 'react-native'

export default StyleSheet.create({
  container: {
    paddingTop: 10
  },
  input: {
    fontFamily: Platform.select({
      android: 'monospace',
      ios: 'Courier New'
    })
  }
})
