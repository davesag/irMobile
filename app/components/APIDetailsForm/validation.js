export const validateKey = value => (value ? undefined : 'API Key is Required')

export const validateSecret = value =>
  value ? undefined : 'API Secret is Required'
