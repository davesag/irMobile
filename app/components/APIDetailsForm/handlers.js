import { cleanKey, cleanSecret } from './clean'

const handlers = ({ handleChange, values }) => ({
  handleChangeKey: value => {
    handleChange('apiKey')(cleanKey(value))
  },

  handleChangeSecret: value => {
    handleChange('apiSecret')(cleanSecret(value))
  },

  handleFocusKey: () => {
    if (values.apiKey) handleChange('apiKey')(null)
  },

  handleFocusSecret: () => {
    if (values.apiSecret) handleChange('apiSecret')(null)
  }
})

export default handlers
