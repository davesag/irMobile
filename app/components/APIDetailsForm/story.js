import React from 'react'
import { storiesOf } from '@storybook/react-native'
import { SafeAreaView } from 'react-native'

import APIDetailsForm from '.'

const keys = {
  apiKey: '1234567890',
  apiSecret: 'abcd123'
}

const onClear = () => {
  console.log('Cleared the form')
}

const onSave = () => {
  console.log('Saved the form')
}

storiesOf('APIDetailsForm', module)
  .add('empty', () => (
    <SafeAreaView>
      <APIDetailsForm onSave={onSave} onClear={onClear} />
    </SafeAreaView>
  ))
  .add('with values', () => (
    <SafeAreaView>
      <APIDetailsForm
        keys={keys}
        requireAuth={true}
        onSave={onSave}
        onClear={onClear}
      />
    </SafeAreaView>
  ))
  .add('saving', () => (
    <SafeAreaView>
      <APIDetailsForm
        keys={keys}
        requireAuth={true}
        saving
        onSave={onSave}
        onClear={onClear}
      />
    </SafeAreaView>
  ))
