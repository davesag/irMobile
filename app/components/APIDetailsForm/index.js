import React, { Component } from 'react'
import PropTypes from 'prop-types'
import { Keyboard, View } from 'react-native'
import { Button, Switch, TextInput, Text } from 'react-native-paper'

import { keysShape } from '../../shapes'

import styles from './styles'

const cleanKey = key => key.replace(/[^a-f0-9-]/gi, '')
const cleanSecret = key => key.replace(/[^a-f0-9]/gi, '')

class APIDetailsForm extends Component {
  static propTypes = {
    keys: PropTypes.shape(keysShape),
    requireAuth: PropTypes.bool,
    saving: PropTypes.bool,
    onSave: PropTypes.func.isRequired,
    onClear: PropTypes.func.isRequired
  }

  static defaultProps = {
    requireAuth: false,
    keys: null,
    saving: false
  }

  constructor(props) {
    super(props)

    this.state = {
      apiKey: {
        value: props.keys ? props.keys.apiKey : null,
        error: null
      },
      apiSecret: {
        value: props.keys ? props.keys.apiSecret : null,
        error: null
      },
      requireAuth: {
        value: props.requireAuth
      },
      dirty: false
    }
  }

  updateField = field => value => {
    if (value) this.setState({ [field]: { value, error: null }, dirty: true })
    else
      this.setState({
        [field]: { value, error: 'Required field' },
        dirty: true
      })
  }

  toggleField = field => () => {
    this.setState({ [field]: { value: !this.state[field].value }, dirty: true })
    Keyboard.dismiss()
  }

  updateApiKey = text => {
    this.updateField('apiKey')(cleanKey(text))
  }

  updateApiSecret = text => {
    this.updateField('apiSecret')(cleanSecret(text))
  }

  canNotBeSubmitted = () => {
    const { apiKey, apiSecret, dirty } = this.state
    return (
      !dirty ||
      apiKey.errors ||
      apiSecret.errors ||
      !apiKey.value ||
      !apiSecret.value
    )
  }

  submit = () => {
    const { apiKey, apiSecret, requireAuth } = this.state
    this.setState({ dirty: false })
    this.props.onSave({
      apiKey: apiKey.value,
      apiSecret: apiSecret.value,
      requireAuth: requireAuth.value
    })
    Keyboard.dismiss()
  }

  clear = () => {
    const { onClear } = this.props
    this.setState({
      apiKey: { value: null, error: null },
      apiSecret: { value: null, error: null },
      requireAuth: { value: false },
      dirty: true
    })
    onClear()
  }

  render() {
    const { apiKey, apiSecret, requireAuth } = this.state
    const { saving } = this.props
    const disabled = this.canNotBeSubmitted()

    return (
      <View style={styles.container}>
        <TextInput
          autoCapitalize="none"
          autoCorrect={false}
          secureTextEntry
          clearTextOnFocus
          inputStyle={styles.input}
          value={apiKey.value}
          label="apiKey"
          error={apiKey.error}
          errorMessage={apiKey.error}
          onChangeText={this.updateApiKey}
          onSubmitEditing={Keyboard.dismiss}
        />
        <TextInput
          autoCapitalize="none"
          autoCorrect={false}
          secureTextEntry
          clearTextOnFocus
          inputStyle={styles.input}
          value={apiSecret.value}
          label="apiSecret"
          error={apiSecret.error}
          errorMessage={apiSecret.error}
          onChangeText={this.updateApiSecret}
          onSubmitEditing={Keyboard.dismiss}
        />
        <View
          style={{ marginTop: 10, flexDirection: 'row', alignItems: 'center' }}
        >
          <Switch
            value={requireAuth.value}
            onValueChange={this.toggleField('requireAuth')}
          />
          <Text style={{ marginLeft: 10 }}>Require passcode for each use</Text>
        </View>
        <View style={styles.buttons}>
          <Button
            style={styles.saveButton}
            mode="contained"
            title="Save Keys"
            disabled={disabled || saving}
            loading={saving}
            onPress={this.submit}
            accessibilityLabel="Save Keys"
          >
            Save Keys
          </Button>
          <Button
            style={styles.clearButton}
            mode="outlined"
            title="Clear Keys"
            disabled={saving || (!apiSecret.value && !apiKey.value)}
            onPress={this.clear}
            accessibilityLabel="Clear Keys"
          >
            Clear Keys
          </Button>
        </View>
      </View>
    )
  }
}

export default APIDetailsForm
