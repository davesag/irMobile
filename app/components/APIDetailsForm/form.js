import React from 'react'
import { View } from 'react-native'
import { Button, Switch, TextInput, Text, HelperText } from 'react-native-paper'
import { Field } from 'formik'

import { validateKey, validateSecret } from './validation'
import handlers from './handlers'

import styles from './styles'

const form = ({ saving, onClear }) => ({
  values,
  touched,
  errors,
  dirty,
  handleChange,
  handleBlur,
  handleSubmit
}) => {
  const hasErrors = Object.keys(errors).length !== 0
  const canNotBeSubmitted = hasErrors || !dirty || saving
  const {
    handleChangeKey,
    handleChangeSecret,
    handleFocusKey,
    handleFocusSecret
  } = handlers({ values, handleChange })

  const hasFieldErrors = ['apiKey', 'apiSecret'].reduce((acc, elem) => {
    acc[elem] = Boolean(errors[elem] && touched[elem])
    return acc
  }, {})

  return (
    <View style={styles.container}>
      <View>
        <Field
          name="apiKey"
          style={styles.input}
          autoCapitalize="none"
          autoCorrect={false}
          secureTextEntry
          label="apiKey"
          disabled={saving}
          error={hasFieldErrors.apiKey}
          onChangeText={handleChangeKey}
          onBlur={handleBlur('apiKey')}
          onFocus={handleFocusKey}
          value={values.apiKey}
          component={TextInput}
          validate={validateKey}
        />
        <HelperText type="error" visible={hasFieldErrors.apiKey}>
          {errors.apiKey}
        </HelperText>
      </View>
      <View>
        <Field
          name="apiSecret"
          style={styles.input}
          autoCapitalize="none"
          autoCorrect={false}
          secureTextEntry
          label="apiSecret"
          disabled={saving}
          error={hasFieldErrors.apiSecret}
          onChangeText={handleChangeSecret}
          onBlur={handleBlur('apiSecret')}
          onFocus={handleFocusSecret}
          value={values.apiSecret}
          component={TextInput}
          validate={validateSecret}
        />
        <HelperText type="error" visible={hasFieldErrors.apiSecret}>
          {errors.apiSecret}
        </HelperText>
      </View>
      <View style={styles.switchWrapper}>
        <Switch
          value={values.requireAuth}
          onValueChange={handleChange('requireAuth')}
        />
        <Text style={styles.switchLabel}>Require passcode for each use</Text>
      </View>
      <View style={styles.buttons}>
        <Button
          style={styles.saveButton}
          mode="contained"
          disabled={canNotBeSubmitted}
          onPress={handleSubmit}
          accessibilityLabel="Save Keys"
        >
          Save Keys
        </Button>
        <Button
          style={styles.clearButton}
          mode="outlined"
          onPress={onClear}
          disabled={saving}
          accessibilityLabel="Clear Keys"
        >
          Clear Keys
        </Button>
      </View>
    </View>
  )
}

export default form
