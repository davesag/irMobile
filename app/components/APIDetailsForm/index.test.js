import 'react-native'
import React from 'react'
import renderer from 'react-test-renderer'

import APIDetailsForm from '.'

let tree

describe('basic form', () => {
  beforeAll(() => {
    tree = renderer.create(<APIDetailsForm />)
  })

  it('rendered correctly ', () => {
    expect(tree).toMatchSnapshot()
  })
})
