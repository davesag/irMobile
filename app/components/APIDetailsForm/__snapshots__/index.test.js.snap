// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rendering defaults rendered correctly  1`] = `
<View
  style={
    Object {
      "paddingTop": 10,
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "marginTop": 10,
      }
    }
  />
  <View
    style={
      Object {
        "flexDirection": "row",
        "justifyContent": "flex-end",
      }
    }
  />
</View>
`;

exports[`rendering with keys and saving rendered correctly  1`] = `
<View
  style={
    Object {
      "paddingTop": 10,
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "marginTop": 10,
      }
    }
  />
  <View
    style={
      Object {
        "flexDirection": "row",
        "justifyContent": "flex-end",
      }
    }
  />
</View>
`;

exports[`rendering with keys rendered correctly  1`] = `
<View
  style={
    Object {
      "paddingTop": 10,
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "marginTop": 10,
      }
    }
  />
  <View
    style={
      Object {
        "flexDirection": "row",
        "justifyContent": "flex-end",
      }
    }
  />
</View>
`;
